examples/main/main.cpp

```diff
@@ -851,6 +895,84 @@ int main(int argc, char ** argv) {
+
+                    if (cmd.size() == 2 && cmd[0] == "/readfile") {
+                        const std::string filename = cmd[1];
+                        LOG_DBG("reading file: '%s'\n", filename.c_str());
+                        std::ifstream text_file(filename);
+                        if (!text_file) {
+                            LOG("failed to open file '%s'\n", filename.c_str());
+                            continue;
+                        }
+                        pending_input << text_file.rdbuf() << "\n\n";
+                        LOG("read %zu characters from file\n", (size_t) text_file.tellg());
+                        continue;
+                    } else if (cmd.size() == 2 && cmd[0] == "/savesess") {
+                        const std::string filename = cmd[1];
+                        LOG("save session file: '%s'\n", filename.c_str());
+                        size_t res = llama_state_save_file(ctx, filename.c_str(), embd_inp.data(), n_past);
+                        if (res == 0) {
+                            LOG("failed to save session file '%s'\n", filename.c_str());
+                        }
+                        continue;
+                    } else if (cmd.size() == 2 && cmd[0] == "/loadsess") {
+                        const std::string filename = cmd[1];
+                        LOG("load session file: '%s'\n", filename.c_str());
+                        session_tokens.resize(n_ctx);
+                        size_t n_token_count_out;
+                        size_t res = llama_state_load_file(ctx, filename.c_str(), session_tokens.data(), session_tokens.size(), &n_token_count_out);
+                        if (res == 0) {
+                            LOG("failed to load session file '%s'\n", filename.c_str());
+                        } else {
+                            session_tokens.resize(n_token_count_out);
+                            embd_inp = session_tokens;
+                            n_past = n_token_count_out;
+                            llama_kv_cache_seq_rm(ctx, 0, n_past, -1);
+                            LOG("loaded %zu tokens from session file '%s'\n", n_token_count_out, filename.c_str());
+                        }
+                        continue;
+                    } else if (cmd.size() == 1 && cmd[0] == "/regen") {
+                        if (pos_history.empty()) {
+                            LOG("no previous assistant message to regenerate\n");
+                            continue;
+                        }
+                        int last_n_past = pos_history.back();
+                        int n_tokens_removed = n_past - last_n_past;
+                        llama_kv_cache_seq_rm(ctx, 0, last_n_past, -1);
+                        n_remain += n_tokens_removed;
+                        is_interacting = false;
+                        // we intentionally do not reset the sampling, so new message will be more diverse
+                        continue;
```

Comment:
I remember having problems with regeneration without adjusting `prev`. I ended up capturing the entire ring buffer and just restoring it. I did the same to all other values, capturing state instead of calculating. So far I have stable results over usual 20 regenerations (my usual bulk testing) with the following:
```
        restore_smpl(); // rewinds gsmpl->prev
        llama_kv_cache_seq_rm(ctx, 0, rewind_state.kv_cache_pos, -1);
        embd_inp.erase(embd_inp.begin() + rewind_state.embd_inp_size, embd_inp.end()); // not sure
        n_past = rewind_state.n_past_size;
        n_consumed = rewind_state.n_consumed_size;
```

However, I have to say that there is something missing/extra here: when testing `K-Shift` sampler I noticed that initial logits of the first message is different from all later regenerations. Still not sure what's wrong, but maybe it helps.